drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2149 @@
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_SLEEP);
+	if (ret == -ENOENT) {
+		/* Warn but don't block suspend */
+		LOG_WRN("I3C pinctrl sleep state not available");
+	} else if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int i3c_stm32_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	int err;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		err = i3c_stm32_activate(dev);
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		err = i3c_stm32_suspend(dev);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return err;
+}
+#endif
+
+#ifdef CONFIG_I3C_STM32_DMA
+static int i3c_stm32_dma_stream_config(const struct device *dev,
+				       struct i3c_stm32_dma_stream *dma_stream, uint64_t src_addr,
+				       uint64_t dst_addr)
+{
+	if (dma_stream->dma_dev != NULL) {
+		if (!device_is_ready(dma_stream->dma_dev)) {
+			return -ENODEV;
+		}
+	}
+
+	memset(&dma_stream->blk_cfg, 0, sizeof(dma_stream->blk_cfg));
+
+	dma_stream->blk_cfg.source_address = src_addr;
+
+	dma_stream->blk_cfg.dest_address = dst_addr;
+
+	if (dma_stream->src_addr_increment) {
+		dma_stream->blk_cfg.source_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+	} else {
+		dma_stream->blk_cfg.source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+	}
+
+	if (dma_stream->dst_addr_increment) {
+		dma_stream->blk_cfg.dest_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+	} else {
+		dma_stream->blk_cfg.dest_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+	}
+
+	dma_stream->blk_cfg.source_reload_en = 0;
+	dma_stream->blk_cfg.dest_reload_en = 0;
+	dma_stream->blk_cfg.fifo_mode_control = dma_stream->fifo_threshold;
+
+	dma_stream->dma_cfg.head_block = &dma_stream->blk_cfg;
+	dma_stream->dma_cfg.user_data = (void *)dev;
+
+	return 0;
+}
+
+/* Initializes the I3C DMA */
+static int i3c_stm32_init_dma(const struct device *dev)
+{
+	struct i3c_stm32_data *data = dev->data;
+	int err;
+	const struct i3c_stm32_config *config = dev->config;
+	I3C_TypeDef *i3c = config->i3c;
+
+	/*Configure DMA RX */
+	err = i3c_stm32_dma_stream_config(
+		dev, &data->dma_rx, LL_I3C_DMA_GetRegAddr(i3c, LL_I3C_DMA_REG_DATA_RECEIVE_BYTE),
+		0);
+	if (err != 0) {
+		return err;
+	}
+
+	/*Configure DMA RS */
+	err = i3c_stm32_dma_stream_config(dev, &data->dma_rs,
+					  LL_I3C_DMA_GetRegAddr(i3c, LL_I3C_DMA_REG_STATUS), 0);
+	if (err != 0) {
+		return err;
+	}
+
+	/*Configure DMA TX */
+	err = i3c_stm32_dma_stream_config(
+		dev, &data->dma_tx, 0,
+		LL_I3C_DMA_GetRegAddr(i3c, LL_I3C_DMA_REG_DATA_TRANSMIT_BYTE));
+	if (err != 0) {
+		return err;
+	}
+
+	/*Configure DMA TC */
+	err = i3c_stm32_dma_stream_config(dev, &data->dma_tc, 0,
+					  LL_I3C_DMA_GetRegAddr(i3c, LL_I3C_DMA_REG_CONTROL));
+	if (err != 0) {
+		return err;
+	}
+
+	return err;
+}
+#endif
+
+static void i3c_stm32_controller_init(const struct device *dev)
+{
+	struct i3c_stm32_data *data = dev->data;
+	const struct i3c_stm32_config *config = dev->config;
+	I3C_TypeDef *i3c = config->i3c;
+
+#ifdef CONFIG_I3C_USE_IBI
+	LL_I3C_EnableHJAck(i3c);
```

Comment:
you may want to enable HJ Acks after the `i3c_bus_init` in the init function, otherwise there could be device request an assignment before the bus is ready